services/gitlab/gitlab-contributors.service.js

```diff
@@ -0,0 +1,73 @@
+import Joi from 'joi'
+import { optionalUrl } from '../validators.js'
+import { renderContributorBadge } from '../contributor-count.js'
+import GitLabBase from './gitlab-base.js'
+
+const queryParamSchema = Joi.object({
+  gitlab_url: optionalUrl,
+}).required()
+
+const documentation = `
+<p>
+  You may use your GitLab Project Id (e.g. 278964) or your Project Path (e.g. gitlab-org/gitlab )
+</p>
+`
+
+const customDocumentation = `
+<p>
+  Note that only network-accessible jihulab.com and other self-hosted GitLab instances are supported.
+  You may use your GitLab Project Id (e.g. 13953) or your Project Path (e.g. gitlab-cn/gitlab ) in <a href="https://jihulab.com">https://jihulab.com</a>
+</p>
+`
+
+export default class GitlabContributors extends GitLabBase {
+  static category = 'activity'
+  static route = {
+    base: 'gitlab/v/contributor',
+    pattern: ':project+',
+    queryParamSchema,
+  }
+
+  static examples = [
+    {
+      title: 'GitLab contributors',
+      namedParams: {
+        project: 'gitlab-org/gitlab',
+      },
+      staticPreview: this.render({ contributorCount: 418 }),
+      documentation,
+    },
+    {
+      title: 'GitLab (custom server) contributors',
+      queryParams: { gitlab_url: 'https://jihulab.com' },
+      namedParams: {
+        project: 'gitlab-cn/gitlab',
+      },
+      staticPreview: this.render({ contributorCount: 415 }),
+      documentation: customDocumentation,
+    },
+  ]
+
+  static defaultBadgeData = { label: 'contributors' }
+
+  static render({ contributorCount }) {
+    return renderContributorBadge({ contributorCount })
+  }
+
+  async handle({ project }, { gitlab_url: baseUrl = 'https://gitlab.com' }) {
+    // https://docs.gitlab.com/ee/api/repositories.html#contributors
+    const { res } = await this._request({
+      url: `${baseUrl}/api/v4/projects/${encodeURIComponent(
+        project
+      )}/repository/contributors`,
+      options: { searchParams: { page: '1', per_page: '1' } },
+      errorMessages: {
+        404: 'project not found',
+      },
+    })
+    // The total number of contributors is in the `x-total` field in the headers.
+    // https://docs.gitlab.com/ee/api/index.html#other-pagination-headers
+    const contributorCount = res.headers['x-total']
```

Comment:
Even though we are extracting the value from the header instead of the response body we should still validate the value exists and makes sense before we pass the value to `renderContributorBadge()`. You can either do it by applying a Joi schema, or do some checks inline here and manually throw an `InvalidResponse` if needed https://github.com/badges/shields/blob/c73072deed8eb6a5b9f51a06377f5f396ef21842/core/base-service/errors.js#L80

Reply:
done